Support border/margin/padding on mfrac and mspace elements

fred-wang opened the pull request:
#### 500c94501f7f041167e2671087e51df2f2672880
<pre>
Support border/margin/padding on mfrac and mspace elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=276218">https://bugs.webkit.org/show_bug.cgi?id=276218</a>

Reviewed by Rob Buis.

This implements support for border/margin/padding on the mfrac and
mspace elements, following the rules from the MathML Core specification:

- When handling boxes of children during math layout, we consider *margin*
  boxes. Following other renderers, `recomputeLogicalWidth()` is called
  during the layout of children to set the inline margins while
  `computeAndSetBlockDirectionMargins()` is called during the parent
  layout to set the block margins.

- Current math layout is modified so that padding/border are added in
  order to obtain the border box. In the case of mfrac, the fraction
  bar continues to be aligned with the math axis and to cover the
  numerator/denominator&apos;s margin boxes. In the case of mspace, the
  baseline alignment is preserved.

- In bug 262789, a padding start/end of 1px for mfrac was added to the
  UA stylesheet. With the present patch, we take this into account for
  layout, allowing to address the use case mentioned in the spec
  &quot;To avoid visual confusion between the fraction bar and another
   adjacent items (e.g. minus sign or another fraction&apos;s bar) a
   default 1-pixel space is added around the element.&quot;.

Canonical link: <a href="https://commits.webkit.org/280665@main">https://commits.webkit.org/280665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d738c0600fca016f9079cef82f11f80c7cbc3d4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57257 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36585 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9732 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60879 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7700 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59385 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44209 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7890 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46355 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5423 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59287 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34330 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49439 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27218 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31112 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6753 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6705 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7023 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62558 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1170 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7115 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53616 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1175 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49476 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53690 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12658 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/986 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32414 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33499 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34584 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33245 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/mathml/RenderMathMLFraction.cpp:
```diff
@@ -185,21 +185,23 @@ void RenderMathMLFraction::computePreferredLogicalWidths()
-        LayoutUnit denominatorWidth = denominator().maxPreferredLogicalWidth();
-        m_minPreferredLogicalWidth = m_maxPreferredLogicalWidth = std::max(numeratorWidth, denominatorWidth);
+        LayoutUnit numeratorWidth = numerator().maxPreferredLogicalWidth() + numerator().marginLogicalWidth();
+        LayoutUnit denominatorWidth = denominator().maxPreferredLogicalWidth()  + denominator().marginLogicalWidth();
```
Looks like there is an extra ' '.

rwlbuis reviewed (approved):
LGTM.

fred-wang committed:
Support border/margin/padding on mfrac and mspace elements
https://bugs.webkit.org/show_bug.cgi?id=276218

Reviewed by Rob Buis.

This implements support for border/margin/padding on the mfrac and
mspace elements, following the rules from the MathML Core specification:

- When handling boxes of children during math layout, we consider *margin*
  boxes. Following other renderers, `recomputeLogicalWidth()` is called
  during the layout of children to set the inline margins while
  `computeAndSetBlockDirectionMargins()` is called during the parent
  layout to set the block margins.

- Current math layout is modified so that padding/border are added in
  order to obtain the border box. In the case of mfrac, the fraction
  bar continues to be aligned with the math axis and to cover the
  numerator/denominator's margin boxes. In the case of mspace, the
  baseline alignment is preserved.

- In bug 262789, a padding start/end of 1px for mfrac was added to the
  UA stylesheet. With the present patch, we take this into account for
  layout, allowing to address the use case mentioned in the spec
  "To avoid visual confusion between the fraction bar and another
   adjacent items (e.g. minus sign or another fraction's bar) a
   default 1-pixel space is added around the element.".

Canonical link: https://commits.webkit.org/280665@main
